### PR TITLE
New Pika Fathead

### LINKED ISSUE #446: Create New Pika Fathead
# Pika Fathead

Contribute to DuckDuckGo!

screen shot 2016-10-09 at 8 57 29 pm
### What you will need to know?

Either Python (Preferrable for this mission), Ruby, Perl or JavaScript for Fatheads. You will download the Python redis client source and create an **output.txt** file containing the parsed documentation. Read more about Fatheads here: 
https://docs.duckduckhack.com/resources/fathead-overview.html
### The source :rocket:

https://pika.readthedocs.io/en/0.10.0/modules/index.html
### The goal

Help _improve Python language search results_ on DuckDuckGo as part of our Programming Mission! We're trying to reach 100% Instant Answer (IA) coverage for Python-related searches by creating new Instant Answers or improving existing ones.

Some relevent examples:
- [Django Built-ins](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/django_built_ins)
- [Redis](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/redis_commands)
- [Python documentation](https://github.com/duckduckgo/zeroclickinfo-fathead/tree/master/lib/fathead/python)

**Fathead Example (Django Built-ins)**

screen shot 2016-10-09 at 8 59 24 pm
### The steps
- [x] Claim this issue (Comment Below)
- [ ] Please review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-fathead/blob/master/CONTRIBUTING.md)
- [ ] Set up codio environment (See: https://docs.duckduckhack.com/welcome/setup-dev-environment.html) and fork this repository
- [ ] Create the Fathead
- [ ] Create a PR
- [ ] Ping @pjhampton for a review
### Questions?
- [Documentation](https://docs.duckduckhack.com/)
- [The Forum](https://forum.duckduckhack.com/c/programming/python)
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) 
- Ping one of the [Community Leaders](https://duck.co/help/community/community-leaders) or @pjhampton

### PR DESCRIPTION
**Pull request by sriranganathan**
## Type of Change

- [X] New Instant Answer
- [ ] Improvement
  - [ ] Bug fix
  - [ ] Enhancement
- [ ] Non–Instant Answer
  - [ ] Other (Role, Template, Test, Documentation, etc.)
## Description of new Instant Answer, or changes

Added Fathead for Python-Pika using Read the Docs as Source
## Related Issues and Discussions

Fixes #446 
## People to notify

cc @pjhampton 

IA Page: https://duck.co/ia/view/pika

Forum Topic: https://forum.duckduckhack.com/t/new-pika-fathead/1006

### CODE REVIEW

**Comment by daxtheduck** [comment]
@sriranganathan Thanks for the pull request! Unfortunately, we're only accepting contributions and improvements to a limited portion of the search space.
            
The staff and community are currently focused on making DuckDuckGo the [best search engine for programmers](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53). To ensure we're making as much progress as possible, **we are only accepting Instant Answers and updates related to the programming mission**.
    
If this is not related to the Programming Mission, it will likely be closed but you can:
    
1. Help us prevent further confusion by telling us where you found out about DuckDuckHack or where we could've alerted you earlier.
2. Keep this code handy! We'll be expanding into other search topics in the future!
3. Join us at our new home: https://forum.duckduckhack.com/

If this is related to the Programming Mission, it will be reviewed by staff and the community as soon as possible!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by moollaza** [comment]
@sriranganathan thanks, we'll take a look shortly 👍

File: lib/fathead/pika/README.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Pika Fathead
+
+Dependencies:
+  - Python 2.7
+  - BeautifulSoup 4 (pip install beautifulsoup4)
```
Comment (moollaza) [inline lib/fathead/pika/README.txt]: @sriranganathan please make a `requirements.txt` file for this

Comment (sriranganathan) [inline lib/fathead/pika/README.txt]: @moollaza made the changes

**Comment by moollaza** [comment]
@sriranganathan this is looking much better. I will deploy it to Beta momentarily. Can you also commit the generated output.txt? It's small enough that we can track it in the repo 👍

**Comment by moollaza** [comment]
Having some trouble getting this Fathead to work on Beta. It's an internal problem though. Will keep trying

File: lib/fathead/pika/parse.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+
+def build_abstract(info, example=None, detailed_info={}):
+    """
+    Constructs the abstract of an entry according to
+    the new fathead HTML template elements.
+    """
+    abstract = ''
+    if info:
+        abstract += '<p>%s</p' % info.replace('\n', '\\n')
```
Comment (amarlearning) [inline lib/fathead/pika/parse.py]: @sriranganathan `abstract += '<p>%s</p' % info.replace('\n', '\\n')`

The `<p>` is not closing correctly!

Comment (sriranganathan) [inline lib/fathead/pika/parse.py]: @amarlearning thanks for pointing out the bug. Fixed it.

**Comment by moollaza** [comment]
@sriranganathan this is now working!

https://moollaza.duckduckgo.com/?q=TwistedConnection.add_on_connection_unblocked_callback&t=h_&ia=about&iax=1

**Comment by daxtheduck** [comment]
@sriranganathan this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

**Comment by tagawa** [comment]
Hi @sriranganathan. Just wanted to let you know this is now live on DuckDuckGo!
https://duckduckgo.com/?q=Pika+Connection&ia=about

You can also see a chart showing coverage and engagement in the forum topic:
https://forum.duckduckhack.com/t/new-pika-fathead/1006

Thanks again for creating this for Python developers 😃

**Pull request merged**